service-config/src/main/java/com/palantir/config/service/ServiceDiscoveryConfiguration.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2016 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.config.service;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import com.palantir.remoting.ssl.SslConfiguration;
+import com.palantir.tokens.auth.BearerToken;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import org.immutables.value.Value.Immutable;
+import org.immutables.value.Value.Lazy;
+import org.immutables.value.Value.Style;
+
+/**
+ * Configuration class that contains a map of {@code serviceName}s and their respective {@link ServiceConfiguration}s
+ * and provides utility methods for apps to build service clients.
+ */
+@Immutable
+@JsonDeserialize(as = ImmutableServiceDiscoveryConfiguration.class)
+@Style(get = "*", visibility = Style.ImplementationVisibility.PACKAGE)
+public abstract class ServiceDiscoveryConfiguration {
+
+    /**
+     * Fallback API token to be used if the service specific API token is not defined in the
+     * {@link ServiceConfiguration}.
+     */
+    @JsonProperty("apiToken")
+    public abstract Optional<BearerToken> getDefaultApiToken();
+
+    /**
+     * Fallback SSL Configuration to be used if the service specific SSL configuration is not defined in the
+     * {@link ServiceConfiguration}.
+     */
+    @JsonProperty("security")
+    public abstract Optional<SslConfiguration> getDefaultSecurity();
+
+    @JsonProperty("services")
+    abstract Map<String, ServiceConfiguration> originalServices();
+
+    /**
+     * A map of {@code serviceName}s as the keys and their respective {@link ServiceConfiguration}s as the values.
+     */
+    @Lazy
+    @SuppressWarnings("checkstyle:designforextension")
+    public Map<String, ServiceConfiguration> getServices() {
```

Comment:
There should be a test in the serviceDiscoveryConfigurationTests already

Sent from my iPhone

On Apr 14, 2016, at 7:46 PM, Robert Fink <notifications@github.com<mailto:notifications@github.com>> wrote:


In service-config/src/main/java/com/palantir/config/service/ServiceDiscoveryConfiguration.java<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_palantir_http-2Dremoting_pull_96-23discussion-5Fr59809953&d=CwMCaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=2LK6Jot8d42Knk04ei1_FPh1f_GvosvCO-13ihGNoi8&m=kiU3Oy4nbMYyCIIqb3n-SWDN0IdlynMFSU4Xb9-Q0vk&s=aAIblTFesqfago_sC_KIP4KF_pUhomjxfyJDrXZ5_V0&e=>:

> +    /**
> +     * Fallback SSL Configuration to be used if the service specific SSL configuration is not defined in the
> +     * {@link ServiceConfiguration}.
> +     */
> +    @JsonProperty("security")
> +    public abstract Optional<SslConfiguration> getDefaultSecurity();
> +
> +    @JsonProperty("services")
> +    abstract Map<String, ServiceConfiguration> originalServices();
> +
> +    /**
> +     * A map of {@code serviceName}s as the keys and their respective {@link ServiceConfiguration}s as the values.
> +     */
> +    @Lazy
> +    @SuppressWarnings("checkstyle:designforextension")
> +    public Map<String, ServiceConfiguration> getServices() {


last request: can you add a test verifying that getServices().get("foo").getApiToken() respects fallbacks?

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_palantir_http-2Dremoting_pull_96_files_61da0033504c8954e15c09743b14facc2f3d011d-23r59809953&d=CwMCaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=2LK6Jot8d42Knk04ei1_FPh1f_GvosvCO-13ihGNoi8&m=kiU3Oy4nbMYyCIIqb3n-SWDN0IdlynMFSU4Xb9-Q0vk&s=qSPFn5S8Mh9Ap3YaSBkI9ASb15-_a9krfpaEWr77ars&e=>